### PR TITLE
feat(process): implement miranum process-core for c7/8

### PR DESCRIPTION
**Pull request by rami-nk**
**Description**
The process use case is now implemented for Camunda 7 and 8. For testing purposes there now is a process-example which provides a REST-API to start a process. The process-example is going to be replaced by a more useful example in future.

**Checklist**

* Code is easy to understand
* No obvious errors or bugs
* Tests were implemented

### CODE REVIEW

**Comment by hauptmedia** [comment]
Just fyi, the current Cloudevents Spec for this command is here: https://github.com/backoffice-plus/cloudevents-schemas/blob/main/docs/bpmn/command/CreateProcessInstanceRequest.schema.md

I think it would make sense if the REST Api uses the same schema for this? We can discuss it in the next weekly next Wednesday

**Comment by DaAnda97** [comment]
Thank you Julian. We'll discuss that.

File: connect/connect-c7/process-adapter-in-c7/src/test/java/io/miragon/miranum/connect/adapter/in/c7/process/Camunda7ProcessAdapterTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    private ProcessDefinitionApi processDefinitionApi;
+    @Mock
+    private Camunda7BaseVariableValueMapper baseVariableMapper;
+    @InjectMocks
+    private Camunda7ProcessAdapter processAdapter;
+
+    @Test
+    public void startProcess_withValidCommand_startsProcessInstance() throws ApiException {
+        var processKey = "processKey";
+        Map<String, Object> variables = Map.of("var1", "value1", "var2", "value2");
```
Comment (DaAnda97) [inline connect/connect-c7/process-adapter-in-c7/src/test/java/io/miragon/miranum/connect/adapter/in/c7/process/Camunda7ProcessAdapterTest.java:36]: I'd use Java 9 syntax here:

``` java
Map<String, String> test2 = Map.ofEntries(
    entry("a", "b"),
    entry("c", "d")
);
```

File: connect/connect-c7/process-adapter-in-c7/src/test/java/io/miragon/miranum/connect/adapter/in/c7/process/Camunda7ProcessAdapterTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        var processInstanceId = "processInstanceId";
+
+        var startProcessCommand = new StartProcessCommand(processKey, variables);
+        var variableValueDtos = Map.of("var1", new VariableValueDto(), "var2", new VariableValueDto());
```
Comment (DaAnda97) [inline connect/connect-c7/process-adapter-in-c7/src/test/java/io/miragon/miranum/connect/adapter/in/c7/process/Camunda7ProcessAdapterTest.java:40]: If you'd like to reference the keys you defined above I'd define them as constants and use those:

`final String KEY_VARIABLE_1 = "var1"`

File: connect/connect-c7/utils-c7/src/test/java/io/miragon/miranum/connect/c7/utils/Camunda7BaseVariableValueMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class Camunda7BaseVariableValueMapperTest {
+
+    private final Camunda7BaseVariableValueMapper mapper = new Camunda7BaseVariableValueMapper();
+
+    @Test
+    public void testMap_ShouldReturnMappedVariables() {
+        // Given
+        Map<String, Object> variables = new HashMap<>();
```
Comment (DaAnda97) [inline connect/connect-c7/utils-c7/src/test/java/io/miragon/miranum/connect/c7/utils/Camunda7BaseVariableValueMapperTest.java:20]: use same Map definition at all places please

File: connect/connect-c7/utils-c7/src/test/java/io/miragon/miranum/connect/c7/utils/Camunda7BaseVariableValueMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+        // When
+        Map<String, VariableValueDto> result = mapper.map(variables);
+
+        // Then
+        assertTrue(result.isEmpty());
+    }
+
+    @Test
+    public void testCreateValue_ShouldReturnCorrectValue() {
```
Comment (DaAnda97) [inline connect/connect-c7/utils-c7/src/test/java/io/miragon/miranum/connect/c7/utils/Camunda7BaseVariableValueMapperTest.java:46]: A good pattern for tests is to test the edge cases, like null or empty. I can also recommend using ParameterizedTests and check for MAX_INTEGER f.ex.

https://www.baeldung.com/parameterized-tests-junit-5

File: connect/connect-c8/process-adapter-in-c8/src/main/java/io/miragon/miranum/connect/adapter/in/c8/process/Camunda8ProcessAdapter.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    private final ZeebeClient zeebeClient;
+
+    @Override
+    public void startProcess(StartProcessCommand startProcessCommand) throws ProcessStartingException {
+        zeebeClient.newCreateInstanceCommand()
+                .bpmnProcessId(startProcessCommand.getProcessKey()).latestVersion()
+                .variables(startProcessCommand.getVariables())
+                .send()
+                .whenComplete((result, exception) -> {
+                    if (exception == null) {
```
Comment (DaAnda97) [inline connect/connect-c8/process-adapter-in-c8/src/main/java/io/miragon/miranum/connect/adapter/in/c8/process/Camunda8ProcessAdapter.java:23]: you could use `Objects.isNull(obj)` here or its static method `isNull(obj)`

File: examples/process-example/process-example-camunda-7/src/main/resources/application.properties [hunk 6]
```diff
@@ -0,0 +1,3 @@
+server.port=5050
+camunda.bpm.client.base-url=http://localhost:8080/engine-rest
+camunda.autoDeploy.enabled: false
```
Comment (DaAnda97) [inline examples/process-example/process-example-camunda-7/src/main/resources/application.properties:3]: I think that variable has no influence on anything since it's wrong formated => just delete it

**Review by DaAnda97 — APPROVED** [review]
Nice work! Please consider refactoring the points I mentioned before merging :)

File: connect/connect-c7/utils-c7/src/test/java/io/miragon/miranum/connect/c7/utils/Camunda7BaseVariableValueMapperTest.java (hunk 3 above)
Comment (rami-nk) [reply connect/connect-c7/utils-c7/src/test/java/io/miragon/miranum/connect/c7/utils/Camunda7BaseVariableValueMapperTest.java:20]: I would not extract any test data for different modules. Especially if we do not have that many tests. If we have more tests that use that some time in the future we can think about implementing a TestDataSeeder.

Comment (DaAnda97) [reply connect/connect-c7/utils-c7/src/test/java/io/miragon/miranum/connect/c7/utils/Camunda7BaseVariableValueMapperTest.java:20]: I was refering to the syntax. Sry for the misleading formulation.

File: examples/process-example/process-example-camunda-7/src/main/resources/application.properties (hunk 6 above)
Comment (rami-nk) [reply examples/process-example/process-example-camunda-7/src/main/resources/application.properties:3]: It actually does exactly what it says. You can checkout the example of Camunda community: https://github.com/camunda-community-hub/camunda-platform-7-rest-client-java.
Looks weird I know, but works.

Comment (rami-nk) [reply examples/process-example/process-example-camunda-7/src/main/resources/application.properties:3]: More concrete, this file: https://github.com/berndruecker/camunda-7-remote-spring-boot-example/blob/main/src/main/resources/application.properties

File: examples/README.md [hunk 7]
```diff
@@ -4,3 +4,4 @@
 *Follow the links below to browse the examples:*
```
[the remark this answers is not in the document]
Comment (rami-nk) [reply examples/README.md:4]: Was not part of this PR but ok. The examples are going to be different and will need different steps to execute the application or see behavior. For me it is ok to leave it like it is

File: examples/process-example/process-example-camunda-7/src/main/resources/application.properties (hunk 6 above)
Comment (DaAnda97) [reply examples/process-example/process-example-camunda-7/src/main/resources/application.properties:3]: Ok thanks. Could you add that as a comment, please?

**Comment by rami-nk** [comment]
> Just fyi, the current Cloudevents Spec for this command is here: https://github.com/backoffice-plus/cloudevents-schemas/blob/main/docs/bpmn/command/CreateProcessInstanceRequest.schema.md
> 
> I think it would make sense if the REST Api uses the same schema for this? We can discuss it in the next weekly next Wednesday

Thank you Julian. We will definitely discuss it next Week! The REST-API was just for testing purposes and did not intend to be anything we want to use in future.
